### PR TITLE
AEA-365 Update docs. Q&A

### LINKED ISSUE #708: Review and improve documentation
**Is your feature request related to a problem? Please describe.**
Currently, the documentation might not be in line with the code base.

**Describe the solution you'd like**
Read each section and ensure:
- [x] it works
- [x] it is correct relative to the code base
- [x] it is comprehensive enough to cover core concepts

### LINKED ISSUE #742: Provide answers to common questions
**Is your feature request related to a problem? Please describe.**
Some common questions are unanswered by the docs.

**Describe the solution you'd like**
Add a Questions & answer sections which guides to the places in the docs which answer these questions or answers them directly:
- [x] Getting started: understanding messaging, understanding context with other agents
- [x] How does an AEA use blockchain & OEF > show also diagrams
- [x] How does one connect a DB?
- [x] How does one connect a life stream of data?
- [x] How does one connect a frontend? > two ways (connections, server in skills)
- [x] How do agents talk to each other who don’t know each other?

Also provide better guidance to these

### PR DESCRIPTION
**Pull request by Totoual**
## Proposed changes

Adds answers for some common questions.

## Fixes

#708
#727
#728   
#729
#738  1/2 objectives.
#742

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/blockchain-oef-usage.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+The AEA framework enables the agents to interact with the blockchain in the form of transaction completion. Currently, the framework supports
```
Comment (DavidMinarsch) [inline docs/blockchain-oef-usage.md]: `with public blockchains to complete transactions.`

File: docs/blockchain-oef-usage.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+two different networks natively: the `Fetch.ai` network and the `Ethereum` network.
+
+Consider the following use case:
+
+Once the AEAs successfully end the negotiations, the buyer should send the correct amount to the specified wallet address. Then, the provider
```
Comment (DavidMinarsch) [inline docs/blockchain-oef-usage.md]: end a negotiation

Comment (DavidMinarsch) [reply docs/blockchain-oef-usage.md]: to the specified wallet address of the sender.

File: docs/blockchain-oef-usage.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+validates the transaction digest and sends the data to the client.
```
Comment (DavidMinarsch) [inline docs/blockchain-oef-usage.md]: to the buyer.

File: docs/blockchain-oef-usage.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+The limitation of the previous use case is that we should trust the provider that he will send the data upon successful payment.
+
+## Ledger Apis
+
+We interact with the blockchains through a module called `LedgerApis`. The `ledger_apis` module contains implementations of each ledger API object (abstract class) that wraps the libraries of each network. With these implementations, we can create accounts for the specified networks. The account contains the private/public key pair and a wallet address. Also with the account, we will be able to
```
Comment (DavidMinarsch) [inline docs/blockchain-oef-usage.md]: The `LedgerApis` module

Comment (DavidMinarsch) [reply docs/blockchain-oef-usage.md]: of each network, in particular `FetchAILedgerApi` and `EthereumLedgerApi`.

Comment (DavidMinarsch) [reply docs/blockchain-oef-usage.md]: Accounts are not created in the ledger apis..these are created in the crypto modules.

File: docs/blockchain-oef-usage.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+sign a transaction and send it to the network.  
+
+## OEF
```
Comment (DavidMinarsch) [inline docs/blockchain-oef-usage.md]: Move this to the existing section on the oef

File: docs/connect-a-database.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
+The AEA framework uses models that enable us to create custom behaviour for each agent. Moreover, the framework enables us to use third-party libraries
```
Comment (DavidMinarsch) [inline docs/connect-a-database.md]: Isn't that stuff covered in the ORM guide?

File: docs/connect-a-stream-of-data.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+The AEA framework enables us to extend the agent by creating different models based on usage. To connect a
```
Comment (DavidMinarsch) [inline docs/connect-a-stream-of-data.md]: This should be an answer to a question in Q&A, not its own section

File: docs/discovery-and-communication.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+The AEA framework uses the 'Open Economic Framework' (OEF) to be able to search and discover agents but this doesn't
```
Comment (DavidMinarsch) [inline docs/discovery-and-communication.md]: Move this to the OEF section

File: docs/integration.md [hunk 9]
```diff
@@ -10,20 +10,239 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
 The `Wallet` holds instantiation of the abstract `Crypto` base class, in particular `FetchaiCrypto` and `EthereumCrypto`.
 
 The `LedgerApis` holds instantiation of the abstract `LedgerApi` base class, in particular `FetchaiLedgerApi` and `EthereumLedgerApi`.
```
Comment (DavidMinarsch) [inline docs/integration.md]: Why is this discussion separate from `docs/blockchain-oef-usage.md`?

File: docs/integration.md [hunk 10]
```diff
@@ -10,20 +10,239 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+
+        time_stamp = int(time.time())
+        seller = cast(str, seller)
+        client = cast(str, client)
+        aggregate_hash = sha256_hash(
+            b"".join([seller.encode(), client.encode(), time_stamp.to_bytes(32, "big")])
+        )
+
+        return aggregate_hash.hex()
+```
```
Comment (DavidMinarsch) [inline docs/integration.md]: You should discuss the methods.

File: docs/understanding-context.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
+The AEA framework uses each own way of enabling the agents to communicate with each other. Specifically, we are creating an envelope that contains the details
```
Comment (DavidMinarsch) [inline docs/understanding-context.md]: We already have a section on `envelopes`. This should be merged there.

File: mkdocs.yml [hunk 12]
```diff
@@ -67,6 +67,12 @@ nav:
       - "Build an AEA on a Raspberry Pi": 'raspberry-set-up.md'
       - "Integrate with third parties": 'integration.md'
       - Use multiplexer stand-alone: 'multiplexer-standalone.md'
+    - Q&A:
```
Comment (DavidMinarsch) [inline mkdocs.yml]: Q&A should be one long document with questions and answers (to toggle)

File: docs/connect-a-frontend.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
+This demo will discuss the options we would have if we wanted to connect a frontend to the AEA. The following diagram illustrates
```
Comment (DavidMinarsch) [inline docs/connect-a-frontend.md]: `the options we would have if we wanted to connect` > `the options we have to connect`

File: docs/connect-a-frontend.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
+This demo will discuss the options we would have if we wanted to connect a frontend to the AEA. The following diagram illustrates 
+the two options we are going to discuss.
+
+<center>![How to connect frontend to your AEA](assets/http-integration.png)</center> 
+
+## Case 1
+The first option we have is to create a connection that will handle the incoming requests from the rest API. In this scenario,
```
Comment (DavidMinarsch) [inline docs/connect-a-frontend.md]: create a `Connection` package

File: docs/connect-a-frontend.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+the rest API communicates with the AEA and requests are handled from a connection that we developed. The rest API should send envelops to the AEA, then the multiplexer will pass the message to the correct skill.
```
Comment (DavidMinarsch) [inline docs/connect-a-frontend.md]: requests are handled by the `HTTP Connection` package. The rest API should send CRUD requests to the `HTTP Connection` which translates these into `Envelopes` to be consumed by the correct skill.

File: docs/connect-a-frontend.md [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+the two options we are going to discuss.
+
+<center>![How to connect frontend to your AEA](assets/http-integration.png)</center> 
+
+## Case 1
+The first option we have is to create a connection that will handle the incoming requests from the rest API. In this scenario,
+the rest API communicates with the AEA and requests are handled from a connection that we developed. The rest API should send envelops to the AEA, then the multiplexer will pass the message to the correct skill. 
+
+## Case 2
+The other option would be to create a stand-alone `Multiplexer` with an `OEF` connection that will be able to search and discover other AEAs. In this scenario,
```
Comment (DavidMinarsch) [inline docs/connect-a-frontend.md]: with an `OEF` connection. In this scenario, the frontend needs to incorporate a Multiplexer with an OEF connection. Then the OEF messaging relay can be used to send envelopes from the AEA to the frontend.

File: docs/connect-a-frontend.md [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+the frontend should query the OEF for results. Then we can communicate with the AEA by sending envelopes.
+
+## Conclusion
+
+The major difference between these two approaches is that in the first case we should know the IP of the AEA to be able to communicate. In case 2 though, the whole search and discovery are handled by the OEF.
```
Comment (DavidMinarsch) [inline docs/connect-a-frontend.md]: `the whole search and discovery are handled by the OEF.`> `the whole communication is handled by the OEF.`

File: docs/connect-a-frontend.md [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+The other difference is that in the first case we are limited to communicate with one or a list of known AEAs and it is less dynamic since we will have to manually update
```
Comment (DavidMinarsch) [inline docs/connect-a-frontend.md]: I don't get this part.

File: docs/integration.md [hunk 19]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
 The `Wallet` holds instantiation of the abstract `Crypto` base class, in particular `FetchaiCrypto` and `EthereumCrypto`.
 
 The `LedgerApis` holds instantiation of the abstract `LedgerApi` base class, in particular `FetchaiLedgerApi` and `EthereumLedgerApi`.
+You can think the `Ledger Api` implementation of each blockchain as a wrapper of their python library.
```
Comment (DavidMinarsch) [inline docs/integration.md]: You can think the concrete implementations of the base class `LedgerApi` as wrappers of the blockchain specific python SDK.

File: docs/integration.md [hunk 20]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+    @property
+    @abstractmethod
+    def api(self) -> Any:
+        """
+        Get the underlying API object.
+        If there is no such object, return None.
+        """
+```
+The api property can be used for low-level operation with the concrete ledger APIs.
+Enables us to create an object based on the implementation of each ledger support.
```
Comment (DavidMinarsch) [inline docs/integration.md]: Drop line 34?

File: docs/integration.md [hunk 21]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+        """
+        Get the balance of a given account.
+
+        This usually takes the form of a web request to be waited synchronously.
+
+        :param address: the address.
+        :return: the balance.
+        """
+```
+The `get_balance` method will return the amount of tokens we hold for a specific address.
```
Comment (DavidMinarsch) [inline docs/integration.md]: `will return` > `returns`

File: docs/integration.md [hunk 22]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+
+        :param tx_nonce: verifies the authenticity of the tx
+        :param crypto: the crypto object associated to the payer.
+        :param destination_address: the destination address of the payee.
+        :param amount: the amount of wealth to be transferred.
+        :param tx_fee: the transaction fee.
+        :return: tx digest if successful, otherwise None
+        """
+```
+The `send_transaction` is where we must implement the logic for sending a transaction to the ledger we want.
```
Comment (DavidMinarsch) [inline docs/integration.md]: `must implement the logic for sending a transaction to the ledger`

File: docs/integration.md [hunk 23]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+        :param seller: the address of the seller.
+        :param client: the address of the client.
+        :param tx_nonce: the transaction nonce.
+        :param amount: the amount we expect to get from the transaction.
+        :param tx_digest: the transaction digest.
+
+        :return: True if the transaction referenced by the tx_digest matches the terms.
+        """
+```
+The `is_transaction_settled` and `validate_transaction` are two functions that helps us to verify a transaction digest that we received.
```
Comment (DavidMinarsch) [inline docs/integration.md]: drop `that we received`.

File: docs/integration.md [hunk 24]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+    def generate_tx_nonce(self, seller: Address, client: Address) -> str:
+        """
+        Generate a random str message.
+
+        :param seller: the address of the seller.
+        :param client: the address of the client.
+        :return: return the hash in hex.
+        """
+```
+Lastly, we implemented a support function that generates a random hash to help us with security issues. The sender of the funds must include this hash in the transaction
```
Comment (DavidMinarsch) [inline docs/integration.md]: `to help us with verifying the uniqueness of transactions`.

File: docs/integration.md [hunk 25]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+as extra data for the transaction to considered valid.
```
Comment (DavidMinarsch) [inline docs/integration.md]: to be considered valid

File: docs/integration.md [hunk 26]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+
+Moving forward, we are going to discuss the different implementation of `send_transaction` and `validate_transacaction` for the two natively supported ledgers of the framework.
```
Comment (DavidMinarsch) [inline docs/integration.md]: Next, we

File: docs/connect-a-frontend.md (hunk 18 above)
Comment (Totoual) [reply docs/connect-a-frontend.md]: I mean if we connect a rest api we cannot discover new agents. And we must provide a list of agents that we can communicate. On the other hand, with the OEF communication we can get different results based on our search query.

File: docs/integration.md [hunk 27]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+         **kwargs
+     ) -> Optional[str]:
+         """Submit a transaction to the ledger."""
+         tx_digest = self._api.tokens.transfer(
+             crypto.entity, destination_address, amount, tx_fee
+         )
+         self._api.sync(tx_digest)
+         return tx_digest
+```
+As you can see, the implementation for sending a transcation to the Fetch.ai ledger is not that trivial.
```
Comment (DavidMinarsch) [inline docs/integration.md]: `is not that trivial` > `is relatively trivial`.

File: docs/integration.md [hunk 28]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+The only drawback is that we cannot use the tx_nonce yet.
```
Comment (DavidMinarsch) [inline docs/integration.md]: Make this a `note` in a  box, indicating that this is incomplete

File: docs/integration.md [hunk 29]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+         is_valid = (
+             str(tx_contents.from_address) == client
+             and amount == transfers[seller_address]
+         )
+         is_settled = self.is_transaction_settled(tx_digest=tx_digest)
+         result = is_valid and is_settled
+         return result
+```
+Inside the `validate_transcation` we request the contents of the transaction based on the tx_digest we received. We are checking that the address
+of the client is the same as the one that is inside the `from` field of the transaction. Lastly, we are checking that the transaction is settled if both of these return
```
Comment (DavidMinarsch) [inline docs/integration.md]: New sentence from `if both of these`.

File: docs/integration.md [hunk 30]
```diff
@@ -10,20 +10,337 @@ To this end, the framework wraps APIs to interact with the two ledgers and expos
...
+         is_valid = (
+             tx.get("input") == tx_nonce
+             and tx.get("value") == amount
+             and tx.get("from") == client
+             and tx.get("to") == seller
+         )
+         return is_valid
+```
+The `validate_transaction` and `is_transaction_settled` functions help us to check if a transaction digest is valid and is settled. 
+In the Ethereum API, we can pass the tx_nonce, so we can check that it's the same. If it is different, we consider that transaction as no valid. The same happens if any of amount, client address
```
Comment (DavidMinarsch) [inline docs/integration.md]: `tx_nonce` > use"`"

Comment (DavidMinarsch) [reply docs/integration.md]: quote all variables

File: docs/questions-and-answers.md [hunk 31]
```diff
@@ -0,0 +1,63 @@
+<details><summary>What is the Open Economic Framework?</summary>
```
Comment (DavidMinarsch) [inline docs/questions-and-answers.md]: What is the Open Economic Framework (OEF)

File: docs/questions-and-answers.md [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+<details><summary>How does an AEA use blockchain?</summary>
+The AEA framework enables the agents to interact with public blockchains to complete transactions. Currently, the framework supports
+two different networks natively: the `Fetch.ai` network and the `Ethereum` network. 
+<br><br>
+You can read more about the intergration of ledger <a href="/integration/">here</a>
+
+</details>
+
+<details><summary>How does one connect a DB?</summary>
+The AEA framework uses models that enable us to create custom behaviour for each agent. Moreover, the framework enables us to use third-party libraries hosted on PyPI we can directly reference the external dependencies.
```
Comment (DavidMinarsch) [inline docs/questions-and-answers.md]: this section needs splitting up, let's talk in person

File: docs/questions-and-answers.md [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+So if you want to connect a database you have two options. Either create a wrapper that communicates with the database and imported in a Model,
+you can find this implementation in the weather_station package, or use ORM (Object-relational-mapping), you have to implement the logic inside 
+a class that inherits from the Model abstract class.
+<br><br>
+For a detailed example of how to use ORM follow the <a href='/orm-integration-to-generic/'>ORM use case</a>  
+</details>
+
+
+<details><summary>How does one connect a lifestream of data?</summary>
+You can create a wrapper class that communicates with the source and import this class in your strategy class that inherits from the Model abstract class,
```
Comment (DavidMinarsch) [inline docs/questions-and-answers.md]: import this class in your skill

**Pull request merged**